Libraries/Opc.Ua.PubSub/Encoding/JsonDataSetMessage.cs

```diff
@@ -0,0 +1,674 @@
+        {
+            string fieldName = field.FieldMetaData.Name;
+
+            Variant valueToEncode = field.Value.WrappedValue;
+            // The StatusCode.Good value is not encoded cor3ectly then it shall be ommited
+            if (valueToEncode == StatusCodes.Good && m_fieldTypeEncoding != FieldTypeEncodingMask.Variant)
+            {
+                valueToEncode = Variant.Null;
+            }
+            switch (m_fieldTypeEncoding)
+            {
+                case FieldTypeEncodingMask.Variant:
+                    // If the DataSetFieldContentMask results in a Variant representation,
+                    // the field value is encoded as a Variant encoded using the reversible OPC UA JSON Data Encoding
+                    // defined in OPC 10000-6.
+                    encoder.ForceNamespaceUri = false;
+                    encoder.WriteVariant(fieldName, valueToEncode, true);
+                    break;
+
+                case FieldTypeEncodingMask.RawData:
+                    // If the DataSetFieldContentMask results in a RawData representation,
+                    // the field value is a Variant encoded using the non-reversible OPC UA JSON Data Encoding
+                    // defined in OPC 10000-6
+                    encoder.ForceNamespaceUri = true;
+
+                    encoder.WriteVariant(fieldName, valueToEncode, false);
+                    break;
+
+                case FieldTypeEncodingMask.DataValue:
+                    DataValue dataValue = new DataValue();
+
+                    dataValue.WrappedValue = valueToEncode;
+
+                    if ((FieldContentMask & DataSetFieldContentMask.StatusCode) != 0)
+                    {
+                        dataValue.StatusCode = field.Value.StatusCode;
+                    }
+
+                    if ((FieldContentMask & DataSetFieldContentMask.SourceTimestamp) != 0)
+                    {
+                        dataValue.SourceTimestamp = field.Value.SourceTimestamp;
+                    }
+
+                    if ((FieldContentMask & DataSetFieldContentMask.SourcePicoSeconds) != 0)
+                    {
+                        dataValue.SourcePicoseconds = field.Value.SourcePicoseconds;
+                    }
+
+                    if ((FieldContentMask & DataSetFieldContentMask.ServerTimestamp) != 0)
+                    {
+                        dataValue.ServerTimestamp = field.Value.ServerTimestamp;
+                    }
+
+                    if ((FieldContentMask & DataSetFieldContentMask.ServerPicoSeconds) != 0)
+                    {
+                        dataValue.ServerPicoseconds = field.Value.ServerPicoseconds;
+                    }
+
+                    // If the DataSetFieldContentMask results in a DataValue representation,
+                    // the field value is a DataValue encoded using the non-reversible OPC UA JSON Data Encoding
+                    encoder.ForceNamespaceUri = true;
```

Comment:
how is this being handled? is it reset after use?

Reply:
This should be safe. The flag is set for each field and there is single encoding type for a dataSetMessage.
I put a logic to revert to initial value anyway.